tests/cpp/test_learner.cc

```diff
@@ -115,4 +116,118 @@ TEST(Learner, SLOW_CheckMultiBatch) {
+  auto p_dmat = *pp_dmat;
+
+  std::vector<bst_float> labels(kRows);
+  for (size_t i = 0; i < labels.size(); ++i) {
+    labels[i] = i;
+  }
+  p_dmat->Info().labels_.HostVector() = labels;
+  std::vector<std::shared_ptr<DMatrix>> mat {p_dmat};
+
+  std::unique_ptr<Learner> learner {Learner::Create(mat)};
+  learner->Configure({Arg{"predictor", "gpu_predictor"},
+                      Arg{"n_gpus", "-1"}});
+  learner->InitModel();
+  learner->UpdateOneIter(0, p_dmat.get());
+  ASSERT_EQ(learner->GetLearnerTrainParameter().gpu_id, 0);
+  ASSERT_EQ(learner->GetLearnerTrainParameter().n_gpus, -1);
+
+  dmlc::TemporaryDirectory tempdir;
+  const std::string fname = tempdir.path + "/model.bst";
+
+  {
+    // Create a scope to close the stream before next read.
+    std::unique_ptr<dmlc::Stream> fo(dmlc::Stream::Create(fname.c_str(), "w"));
+    learner->Save(fo.get());
+  }
+
+  std::unique_ptr<dmlc::Stream> fi(dmlc::Stream::Create(fname.c_str(), "r"));
+  learner->Load(fi.get());
+  ASSERT_EQ(learner->GetLearnerTrainParameter().gpu_id, 0);
+  ASSERT_EQ(learner->GetLearnerTrainParameter().n_gpus, 0);
+
+  delete pp_dmat;
+}
+
+// Tests for automatic GPU configuration.
+TEST(Learner, GPUConfiguration) {
+  using Arg = std::pair<std::string, std::string>;
+  size_t constexpr kRows = 10;
+  auto pp_dmat = CreateDMatrix(kRows, 10, 0);
+  auto p_dmat = *pp_dmat;
+  std::vector<std::shared_ptr<DMatrix>> mat {p_dmat};
+  std::vector<bst_float> labels(kRows);
+  for (size_t i = 0; i < labels.size(); ++i) {
+    labels[i] = i;
+  }
+  p_dmat->Info().labels_.HostVector() = labels;
+  {
+    std::unique_ptr<Learner> learner {Learner::Create(mat)};
+    learner->Configure({Arg{"booster", "gblinear"},
+                        Arg{"updater", "gpu_coord_descent"}});
+    learner->InitModel();
+    learner->UpdateOneIter(0, p_dmat.get());
+    ASSERT_EQ(learner->GetLearnerTrainParameter().gpu_id, 0);
+    ASSERT_EQ(learner->GetLearnerTrainParameter().n_gpus, 1);
+  }
+  {
+    std::unique_ptr<Learner> learner {Learner::Create(mat)};
+    learner->Configure({Arg{"tree_method", "gpu_exact"}});
+    learner->InitModel();
+    learner->UpdateOneIter(0, p_dmat.get());
+    ASSERT_EQ(learner->GetLearnerTrainParameter().gpu_id, 0);
+    ASSERT_EQ(learner->GetLearnerTrainParameter().n_gpus, 1);
+  }
+  {
+    std::unique_ptr<Learner> learner {Learner::Create(mat)};
+    learner->Configure({Arg{"tree_method", "gpu_hist"}});
+    learner->InitModel();
+    learner->UpdateOneIter(0, p_dmat.get());
+    ASSERT_EQ(learner->GetLearnerTrainParameter().gpu_id, 0);
+    ASSERT_EQ(learner->GetLearnerTrainParameter().n_gpus, 1);
+  }
+  {
+    // with CPU algorithm
+    std::unique_ptr<Learner> learner {Learner::Create(mat)};
+    learner->Configure({Arg{"tree_method", "hist"}});
+    learner->InitModel();
+    learner->UpdateOneIter(0, p_dmat.get());
+    ASSERT_EQ(learner->GetLearnerTrainParameter().gpu_id, 0);
+    ASSERT_EQ(learner->GetLearnerTrainParameter().n_gpus, 0);
```

Comment:
So it's expected that `n_gpus` is automatically set to 0 for CPU algorithms?

Reply:
@hcho3 Yes.  That's the plan, choosing GPU parameters based on algorithms/predictor.